Makeconf

```diff
@@ -1,30 +1,39 @@
-####======= Common Declarations for *all* ESS -*-Makefile-*-s ==========
-
 ## To install ESS for all users on your unix system:
-## 1. Edit Section 1 and 2 below
-## 2. Execute: gmake install
+## 1. Check variables below, edit as necessary
+## 2. make
+## 3. make install
 
 ## Section 1
-## Installation variables for your emacs variant
+## Installation variables for your Emacs variant
 ##
 ## Variable        Description
 ## EMACS           Path to GNU Emacs
-## EMACSBATCH      Emacs command for compiling elisp files
-## SITELISP        Destination of site-lisp; but some variants have no site-lisp
-## LISPDIR         Destination of compiled elisp files
-## INFODIR         Destination of info files
-## ETCDIR          Destination of script and icon files
-## PREFIX(DESTDIR) Directory prepended to SITELISP, INFODIR, DOCDIR & ETCDIR
-##                 Specify either PREFIX or DESTDIR to over-ride /usr
-DESTDIR=/usr
-PREFIX=$(DESTDIR)
+## EMACSBATCH      How to run Emacs as batch
+## SITELISP        Destination of site-lisp
+## ESSDIR          Destination of ESS
+## JULIA_DIR       Path to julia-mode
+DESTDIR ?= /usr
+PREFIX ?= $(DESTDIR)
 
 ##__ GNU Emacs __
 EMACS ?= emacs
-SITELISP=$(PREFIX)/share/emacs/site-lisp
-LISPDIR=$(SITELISP)/ess
-INFODIR=$(PREFIX)/share/info
-ETCDIR =$(PREFIX)/share/emacs/etc/ess
+SITELISP ?= $(PREFIX)/share/emacs/site-lisp
+ESSDIR ?= $(SITELISP)/ess
+EMACSBATCH ?= $(EMACS) -batch -Q -L $(JULIA_DIR)
+
+# Load path:
+ifndef JULIA_DIR
+
+ELPA_DIR ?= ~/.emacs.d/elpa
+
```

Comment:
This will break all those for whom elpa is not there, including myself. 

How about creating local directory .ELPA and fetch julia there if not found? I have a simple [setup](https://github.com/polymode/poly-R/blob/master/targets/melpa.el) in polymode to do so. Only a fraction of that is needed here.

Reply:
Actually we can unify it with the travis dependency fetcher.

 I think relying on global .emacs.d/elpa is not good idea because people can test with multiple version of emacs and will run into byte compilation issues.

Reply:
> This will break all those for whom elpa is not there, including myself.

I don't have julia-mode at ~/.emacs.d/elpa either, it's just what I figured was most common. You can tell the Makefile where julia-mode is at by passing the variable JULIA_DIR. 

I really do not like the idea of accessing the network. I oftentimes work without internet access and it's annoying not to be able to run ESS's tests just because I can't download julia.

Reply:
Well, either way without the network one won't be able to test/compile ESS, am I right? Except that in the proposed solution one has to install julia manually and emacs version conflicts are likely. Do I understand your proposal correctly?

With what I propose one would need to have network access once, as the .ELPA dir is never deleted unless requested.

> You can tell the Makefile where julia-mode is at by passing the variable JULIA_DIR.

Sorry, but this is a no-go AFAIC. Currently one just has to `export EMACS=...` before `make ...`. I don't think we want to complicate that.

Reply:
> Well, either way without the network one won't be able to test/compile ESS, am I right? Except that in the proposed solution one has to install julia manually and emacs version conflicts are likely. Do I understand your proposal correctly?

I don't think so? My solution doesn't require network access, it does assume you've got julia-mode somewhere. 

By Emacs version conflicts, are you talking about issues with loading elc files compiled with e.g. Emacs 26 on Emacs 25? That's possible, I guess. Not sure how likely it is, and we can fix it if/when it becomes an issue.

> With what I propose one would need to have network access once, as the .ELPA dir is never deleted unless requested.

But we also have 0 control over what version of julia-mode is in there. We also have to keep maintaining some mechanism of downloading julia-mode, compiling it, putting it on load-path, etc.

Reply:
> I don't think so? My solution doesn't require network access, it does assume you've got julia-mode somewhere.

This does mean that the user needs network to download julia-mode for the first time. Just like with my proposal. 

>  Not sure how likely it is, and we can fix it if/when it becomes an issue.

Very likely if I installed julia-mode on emacs 27 and then test with emacs 25, which I do all the time. 

> But we also have 0 control over what version of julia-mode is in there

Melpa version. Each emacs version will have a sub-dir in ELPA dir with it's own compiled julia-mode. It's a full-control AFAIC. Big advantage of this is that if we have other dependencies (and we will have for ESS+) then those will be fetched there as well. It's a mini-cask in 20 lines of code if you want. 

> We also have to keep maintaining some mechanism of downloading julia-mode

It's the same mechanism that you use on travis - melpa.

Reply:
> This does mean that the user needs network to download julia-mode for the first time. Just like with my proposal.

No, they could get it from a flash drive or other media.

>> Not sure how likely it is, and we can fix it if/when it becomes an issue.

> Very likely if I installed julia-mode on emacs 27 and then test with emacs 25, which I do all the time.

Isn't this an issue with the current setup, though?

>> But we also have 0 control over what version of julia-mode is in there

> Melpa version. Each emacs version will have a sub-dir in ELPA dir with it's own compiled julia-mode. It's a full-control AFAIC

It's not full control at all. Let's say I'm trying to track down a bug in ess-julia and I want to test against different versions of julia-mode. That's not easy/possible if we just rely on MELPA. 

Using MELPA is also a big security risk since it's downloading and running unreviewed code from the internet. I'd really prefer a solution with at least the option of using a user-specified julia-mode.

Reply:
Arh. We are just crazy.  We just do git submodules for our deps and be done with all the complexities.